packages/loader/container-loader/src/audience.ts

```diff
@@ -3,29 +3,40 @@
  * Licensed under the MIT License.
  */
 
-import { EventEmitter } from "@fluid-internal/client-utils";
-import { IAudienceOwner } from "@fluidframework/container-definitions/internal";
+import { TypedEventEmitter } from "@fluid-internal/client-utils";
+import { IAudienceEvents, IAudienceOwner } from "@fluidframework/container-definitions/internal";
 import { assert } from "@fluidframework/core-utils";
 import { IClient } from "@fluidframework/protocol-definitions";
 
 /**
  * Audience represents all clients connected to the op stream.
  */
-export class Audience extends EventEmitter implements IAudienceOwner {
+export class Audience extends TypedEventEmitter<IAudienceEvents> implements IAudienceOwner {
 	private readonly members = new Map<string, IClient>();
+	private _currentClientId: string | undefined;
 
 	constructor() {
 		super();
 		// We are expecting this class to have many listeners, so we suppress noisy "MaxListenersExceededWarning" logging.
 		super.setMaxListeners(0);
 	}
 
-	public on(
-		event: "addMember" | "removeMember",
-		listener: (clientId: string, client: IClient) => void,
-	): this;
-	public on(event: string, listener: (...args: any[]) => void): this {
-		return super.on(event, listener);
+	public get currentClientId() {
+		return this._currentClientId;
+	}
+
+	public setCurrentClientId(clientId: string | undefined): void {
+		if (this._currentClientId !== clientId) {
+			this._currentClientId = clientId;
+			assert(
+				clientId !== undefined,
+				"undefined could be only initial value, and once it is changed, it is never undefined",
+			);
```

Comment:
Should this be outside the if? Right now a call like `.setCurrentClientId(undefined)` before the client connects will not error out, but I think it might be a better contract if it does, just like it would after the client connects? No reason anyone should ever call it like that, right?

Reply:
Maybe? Today we get here with undefined when container is initializing - we have clientId that could be undefined or not (loading from stashed state). if we change order, caller would need to add extra IF simply to avoid it.
We assert all over the place that clientId is not undefined on connection (including here), so that case is covered as well

Reply:
If we already have calls that make it here with undefined then I think this is fine.